File: packages/ts/src/strategies/fix-inner-quotes.ts
/**
 * fix-inner-quotes: Escape unescaped double-quote characters inside JSON string values.
 * Must-isolate: runs after fix-quotes.
 * Scans the text char-by-char tracking string context.
 */
export const fixInnerQuotes = (text: string): string => {
  let result = "";
  let inString = false;
  let i = 0;

  while (i < text.length) {
    const ch = text[i];
    if (!inString) {
      if (ch === '"') {
        inString = true;
        result += ch;
        i++;
        continue;
      }
      result += ch;
      i++;
      continue;
    }
    // Inside a string
    if (ch === "\\") {
      // Consume escape sequence as-is
      result += ch;
      i++;
      if (i < text.length) {
        result += text[i];
        i++;
      }
      continue;
    }
    if (ch === '"') {
      // Check if this quote closes the string or is embedded
      // Heuristic: if next non-whitespace char is : , } ] then it closes
      let j = i + 1;
      while (j < text.length && (text[j] === " " || text[j] === "\t")) j++;
      const next = text[j];
      const isCloser =
        next === ":" ||
        next === "," ||
        next === "}" ||
        next === "]" ||
        next === undefined ||
        j >= text.length;
      if (isCloser) {
        inString = false;
        result += ch;
      } else {
        result += '\\"';
      }
      i++;
      continue;
    }
    result += ch;
    i++;
  }
  return result;
};

File: packages/python/src/output_guard/strategies/fix_inner_quotes.py
"""fix-inner-quotes: Escape unescaped double-quotes inside JSON string values.

Mirror of packages/ts/src/strategies/fix-inner-quotes.ts.
"""

from __future__ import annotations


def fix_inner_quotes(text: str) -> str:
    out: list[str] = []
    in_string = False
    i = 0
    n = len(text)
    while i < n:
        ch = text[i]
        if not in_string:
            if ch == '"':
                in_string = True
                out.append(ch)
                i += 1
                continue
            out.append(ch)
            i += 1
            continue
        if ch == "\\":
            out.append(ch)
            i += 1
            if i < n:
                out.append(text[i])
                i += 1
            continue
        if ch == '"':
            j = i + 1
            while j < n and text[j] in (" ", "\t"):
                j += 1
            nxt = text[j] if j < n else None
            is_closer = nxt in (":", ",", "}", "]", None) or j >= n
            if is_closer:
                in_string = False
                out.append(ch)
            else:
                out.append('\\"')
            i += 1
            continue
        out.append(ch)
        i += 1
    return "".join(out)
